### PR TITLE
EKF3: Fix range beacon fusion bugs

### PR DESCRIPTION
**Pull request by priseborough**
**BACKGROUND**

The EKF3 uses a multi-hypothesis filter to resolve the height ambiguity when using range beacons that are co-planar in the horizontal (All beacons are at the same or similar height). This filter calculates the vertical position of the range beacons relative to the EKF origin using two hypothesis, one where the beacons are below the flight vehicle and one where they are above. The hypothesis with the most stable vertical position is used. The stability is calculated by low pass filtering the magnitude of the vertical offset changes resulting from each range measurement fusion.

**CHANGES**

This fixes the following bugs:

1) The vertical position calculated by the multi-hypothesis filter was only being updated when the hypothesis selection changed between upper/lower.
2) The hypothesis filer was using range beacon positions that had already had an offset applied which was creating a circular dependancy and reducing stability of the estimate
3) There was an typographical error in the LPF used to filter the upper hypothesis vertical position change.

The following improvements were made:

1) The default LPF filter constant was reduced to allow the hypothesis to change more rapidly so that when a flight vehicle  ascends past the beacon constellation plane, the correct hypothesis can be selected sooner.
2) Spike protection was added to the LPF input.

**TESTING STATUS**

This has only been SITL tested to date using the Copter simulation with the following parameters:

[sitl_bcn.txt](https://github.com/ArduPilot/ardupilot/files/1089923/sitl_bcn.txt)

EKF and truth vertical position

screen shot 2017-06-21 at 9 21 43 am

EKF and beacon receiver horizontal position

screen shot 2017-06-21 at 9 22 55 am

Beacon range innovations

screen shot 2017-06-21 at 9 16 35 am

Beacon vertical position hypothesis 

screen shot 2017-06-21 at 9 17 33 am

### CODE REVIEW

**Comment by priseborough** [comment]
Here is a comparison between the legacy EKF2 and the revised EKF3 range innovations. The changes cannot be simply copied across to EKF2. Further work is required to apply the same fixes to EKF2.

screen shot 2017-06-21 at 10 22 33 am

**Comment by priseborough** [comment]
I was able to get a  successful indoor test using the Pozyx beacons. The raw range data with this setup was very noisy:

screen shot 2017-06-21 at 10 29 08 pm

Range Innovations once the initial alignment completed were zero mean:

screen shot 2017-06-21 at 10 26 25 pm

The beacon location vertical offset plus the beacon height parameter value of 1.5  tracked the upper or lower hypothesis value. The noisy data made it jump around, however the vehicle was close to the beacon vertical position and passed through the horizontal plane defined by the beacon positions so either hypothesis could be selected:

screen shot 2017-06-21 at 10 25 53 pm

The position was stable and tracked the real movement:

screen shot 2017-06-21 at 10 28 23 pm

Height tracked the baro:

screen shot 2017-06-21 at 10 27 35 pm

File: libraries/AP_NavEKF3/AP_NavEKF3_RngBcnFusion.cpp [hunk 1]
```diff
@@ -552,20 +552,21 @@ void NavEKF3_core::CalcRangeBeaconPosDownOffset(float obsVar, Vector3f &vehicleP
...
+        float stateChange = innov * gain;
+        OffsetMaxInnovFilt = (1.0f - filtAlpha) * OffsetMaxInnovFilt + fminf(fabsf(filtAlpha * stateChange) , 1.0f);
+
         // Reject range innovation spikes using a 5-sigma threshold unless aligning
         if ((sq(innov) < sq(innovGateWidth) * innovVar) || aligning) {
-            // calculate the Kalman gain
-            gain = (bcnPosOffsetMaxVar * obsDeriv) / innovVar;
 
             // state update
             bcnPosDownOffsetMax -= innov * gain;
```
Comment (OXINARF) [inline libraries/AP_NavEKF3/AP_NavEKF3_RngBcnFusion.cpp]: `stateChange`? Doesn't make much difference but it is changed below.

**Comment by ptsneves** [comment]
We gave this PR a try with the marvel mind and it didn't go well. How do I attach logs? What information do you need?

**Comment by priseborough** [comment]
@ptsneves  Can you provide a link to the file somewhere in a dropbox or google drive?

Also please describe your test setup, the location of beacons including their height, the location of the vehicle, it's starting orientation and the movement sequence you performed. Log data has to be gathered with LOG_DISARMED=1

**Comment by priseborough** [comment]
@OXINARF I don't want to hold up this PR with the marvelmind issue, because marvelmind integration is still a work in progress.

**Comment by OXINARF** [comment]
Merged, thanks! I've asked @rmackay9 and he said he's fine testing after merge.

@ptsneves Better to open an issue for Marvelmind problems, you can attach logs to GitHub by drag and dropping them in the text box.

**Pull request merged**